File: 111.二叉树的最小深度.ts
/*
 * @lc app=leetcode.cn id=111 lang=typescript
 *
 * [111] 二叉树的最小深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

// class TreeNode {
//     val: number;
//     left: TreeNode | null;
//     right: TreeNode | null;
//     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
//         this.val = val === undefined ? 0 : val;
//         this.left = left === undefined ? null : left;
//         this.right = right === undefined ? null : right;
//     }
// }

function minDepth(root: TreeNode | null): number {
    if (!root) {
        return 0;
    }
    let minHeight = 1;
    let queue = [root];
    while (queue.length) {
        const len = queue.length;
        for (let i = 0; i < len; i += 1) {
            const node = queue.shift();
            if (!node?.left && !node?.right) {
                return minHeight;
            }
            if (node.left) {
                queue.push(node.left);
            }
            if (node.right) {
                queue.push(node.right);
            }
        }
        minHeight += 1;
    }
    return minHeight;
}
// @lc code=end

File: 111.二叉树的最小深度.py
#
# @lc app=leetcode.cn id=111 lang=python3
#
# [111] 二叉树的最小深度
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root:
            return 0

        minH = 1
        arr = [root]
        while len(arr):
            temp = []
            while len(arr):
                node = arr.pop()
                if not node.left and not node.right:
                    return minH
                if node.left:
                    temp.append(node.left)
                if node.right:
                    temp.append(node.right)
            arr = temp
            minH += 1

        return minH


# @lc code=end
